### DISCUSSION TITLE
Does csstree supports nested classes/selectors to parse?

### DISCUSSION THREAD

**Discussion by at1217** [csstree/csstree#186] [Q&A]
Suppose I have `style.css` which looks like the following:
```
.class1 {
    margin: 4px;
    .class2 {
        font-size: 16px;
    }
}
```
Then, I parse the css file by calling the following parse and walk functions:
```
const ast = parse(css, {
                        filename,
                        positions: true,
                        onParseError(error) {
                            errors.push(error);
                        }
});

walk(ast, {
    visit: 'Block',
    enter(node) {
        console.log(node.children.head);
    }
});
```
The ` console.log` gives me the following:
```
{
    prev: null;
    next: {
        prev: [Circular],
        next: null,
        data: {
            type: 'Raw',
            loc: [Object],
            value: 'p {\n   font-size: 16px\n   }'
       }
    },
    data: {
        type: 'Declaration',
        loc: {
            source: 'fixtures/css/style.css',
            start: [Object],
            end: [Object]
        },
        important: false
        property: 'margin',
        value: { type: 'Raw', loc: [Object], value: '1' }
    }
}
```

Is there any way that I could get a data object whose type is 'Declaration' for the nested class2 class?

**Comment by lahmatiy** [comment]
A basic support for CSS Nesting was add in [2.3.0](https://github.com/csstree/csstree/releases/tag/v2.3.0) a month ago. However, CSS Nesting [spec](https://www.w3.org/TR/css-nesting-1/) is still in progress and things are changing. For instance, in latest [draft](https://drafts.csswg.org/css-nesting/) `&` selector became optional in most cases. Moreover,  a new suggestions to syntax appears here and there, e.g. latest [Help choose the syntax for CSS Nesting](https://webkit.org/blog/13607/help-choose-from-options-for-css-nesting-syntax/) from webkit authors.
So, the support of CSS Nesting in CSSTree is a subject to change.

**Comment by thescientist13** [comment]
Looks like support for CSS nesting just recently dropped in FF, so I think that covers all major browsers as of now!  🎉 
https://caniuse.com/?search=css%20nesting
<img width="676" alt="Screenshot 2023-09-04 at 09 06 22" src="https://github.com/csstree/csstree/assets/895923/be3224ac-2036-43f9-9e6d-72c21adfc1c1">

**Comment by DimitrijeD** [comment]
If I may I'd like to expand on this topic. Hopefully I haven't reached wrong conclusions nor missunderstood something about the csstree.

In the current version nested rules (particularly those followed by "Declaration") like in the case @at1217 brought up, 
are still type "Raw".

It's fine. We can always attempt to parse that "Raw" node once again - recursively.
I thought it would never create infinite cycles, but I was wrong.
I found some CSS which had incorrect closing comment "*/" which was marked as "Raw". Parsing it produces same result 
"Stylesheet" with one child "Raw" - "*/" . Then parsing him, produces same result - "Stylesheet" with one child "Raw" - "*/"  -> leading to infinite reqursion.
So perhaps reqursive approach to dealing with nested "Raw" nodes is incorrect. But I dont see any other way. It is relatively straight forward to prevent infinite cycles by comparing Raw node value with parsed Raw node value. If they are equal, theres nothing to parse, skip it, otherwise there is valid CSS in parsed raw node.
I was wondering if there are more cases like this or have I just stumbled on one of those rare edge cases?
I'm a bit vague with exact scenario. I could make a spearate discussion if needed. I don't want deviate too much from this discussion.

Thanks.
